File: examples/visualization/node/plotly_example.mjs
#!/usr/bin/env node
/**
 * Plotly.js visualization over the shared GraphForge karate projection.
 *
 * Builds a Plotly figure JSON and a browser-ready HTML page that loads Plotly
 * from a CDN. No interactive browser is opened in CI. Layout matches the
 * Python Plotly example: deterministic circular coordinates from layout_seed.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { project } from "../shared/projection.mjs";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");

export function seededLayout(nodeIds, seed) {
  const sorted = [...nodeIds].sort((a, b) => a - b);
  const count = sorted.length || 1;
  const positions = new Map();
  sorted.forEach((nodeId, index) => {
    const angle = (2 * Math.PI * index) / count + ((seed % 360) * Math.PI) / 180;
    positions.set(nodeId, [Math.cos(angle), Math.sin(angle)]);
  });
  return positions;
}

export function toPlotlyFigure(projection) {
  const positions = seededLayout(
    projection.nodes.map((node) => node.id),
    projection.layout_seed,
  );

  const edgeX = [];
  const edgeY = [];
  for (const edge of projection.edges) {
    const [x0, y0] = positions.get(edge.source);
    const [x1, y1] = positions.get(edge.target);
    edgeX.push(x0, x1, null);
    edgeY.push(y0, y1, null);
  }

  const nodeX = projection.nodes.map((node) => positions.get(node.id)[0]);
  const nodeY = projection.nodes.map((node) => positions.get(node.id)[1]);
  const labels = projection.nodes.map((node) => node.label);

  return {
    data: [
      {
        type: "scatter",
        x: edgeX,
        y: edgeY,
        mode: "lines",
        line: {
          width: projection.style.edge_width,
          color: projection.style.edge_color,
        },
        hoverinfo: "none",
        name: "edges",
      },
      {
        type: "scatter",
        x: nodeX,
        y: nodeY,
        mode: "markers+text",
        text: labels,
        textposition: "top center",
        marker: {
          size: projection.style.node_size,
          color: projection.style.node_color,
        },
        name: "nodes",
      },
    ],
    layout: {
      title: "Zachary karate club (Plotly.js / GraphForge projection)",
      showlegend: false,
      hovermode: "closest",
      xaxis: { showgrid: false, zeroline: false, showticklabels: false },
      yaxis: { showgrid: false, zeroline: false, showticklabels: false },
    },
  };
}

export function buildHtml(figure) {
  const payload = JSON.stringify(figure);
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <title>Zachary karate club (Plotly.js / GraphForge)</title>
  <script src="https://cdn.plot.ly/plotly-2.35.2.min.js"></script>
  <style>
    html, body, #graph { width: 100%; height: 100%; margin: 0; padding: 0; }
  </style>
</head>
<body>
  <div id="graph"></div>
  <script>
    const figure = ${payload};
    Plotly.newPlot("graph", figure.data, figure.layout, {responsive: true});
  </script>
</body>
</html>
`;
}

export async function run(outputDir = join(ROOT, "output")) {
  const projection = await project();
  const figure = toPlotlyFigure(projection);
  mkdirSync(outputDir, { recursive: true });

  const jsonPath = join(outputDir, "plotly_js_karate.json");
  const htmlPath = join(outputDir, "plotly_js_karate.html");
  writeFileSync(jsonPath, `${JSON.stringify(figure, null, 2)}\n`);
  writeFileSync(htmlPath, buildHtml(figure));
  return [htmlPath, jsonPath];
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  const outputDir = process.argv.includes("--output-dir")
    ? process.argv[process.argv.indexOf("--output-dir") + 1]
    : join(ROOT, "output");
  for (const path of await run(outputDir)) {
    console.log(path);
  }
}

File: examples/visualization/python/plotly_example.py
#!/usr/bin/env python3
"""Plotly visualization over the shared GraphForge karate projection."""

from __future__ import annotations

import argparse
import json
import math
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from shared.projection import project  # noqa: E402


def _seeded_layout(node_ids: list[int], seed: int) -> dict[int, tuple[float, float]]:
    """Deterministic circular layout (Plotly has no graph layout seed of its own)."""
    count = len(node_ids)
    positions: dict[int, tuple[float, float]] = {}
    for index, node_id in enumerate(sorted(node_ids)):
        angle = (2.0 * math.pi * index / count) + (seed % 360) * math.pi / 180.0
        positions[node_id] = (math.cos(angle), math.sin(angle))
    return positions


def build_figure(projection: dict):
    import plotly.graph_objects as go

    positions = _seeded_layout(
        [node["id"] for node in projection["nodes"]],
        projection["layout_seed"],
    )
    edge_x: list[float | None] = []
    edge_y: list[float | None] = []
    for edge in projection["edges"]:
        x0, y0 = positions[edge["source"]]
        x1, y1 = positions[edge["target"]]
        edge_x.extend([x0, x1, None])
        edge_y.extend([y0, y1, None])

    node_x = [positions[node["id"]][0] for node in projection["nodes"]]
    node_y = [positions[node["id"]][1] for node in projection["nodes"]]
    labels = [node["label"] for node in projection["nodes"]]

    fig = go.Figure(
        data=[
            go.Scatter(
                x=edge_x,
                y=edge_y,
                mode="lines",
                line={
                    "width": projection["style"]["edge_width"],
                    "color": projection["style"]["edge_color"],
                },
                hoverinfo="none",
                name="edges",
            ),
            go.Scatter(
                x=node_x,
                y=node_y,
                mode="markers+text",
                text=labels,
                textposition="top center",
                marker={
                    "size": projection["style"]["node_size"],
                    "color": projection["style"]["node_color"],
                },
                name="nodes",
            ),
        ],
        layout=go.Layout(
            title="Zachary karate club (Plotly / GraphForge projection)",
            showlegend=False,
            hovermode="closest",
            xaxis={"showgrid": False, "zeroline": False, "showticklabels": False},
            yaxis={"showgrid": False, "zeroline": False, "showticklabels": False},
        ),
    )
    return fig


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=ROOT / "output",
        help="Directory for HTML/JSON artifacts",
    )
    parser.add_argument(
        "--show",
        action="store_true",
        help="Open the interactive figure in a browser (local use only)",
    )
    args = parser.parse_args()

    projection = project()
    fig = build_figure(projection)

    args.output_dir.mkdir(parents=True, exist_ok=True)
    html_path = args.output_dir / "plotly_karate.html"
    json_path = args.output_dir / "plotly_karate.json"
    fig.write_html(str(html_path), include_plotlyjs="cdn", full_html=True)
    json_path.write_text(json.dumps(fig.to_plotly_json(), indent=2) + "\n", encoding="utf-8")
    print(html_path)
    print(json_path)
    if args.show:
        fig.show()


if __name__ == "__main__":
    main()
